### PR TITLE
Update the `maxUnavailable` number of nodes during parallel upgrades

### PR DESCRIPTION
**Pull request by nikimanoledaki**
### Description

This PR's base is #3828.

Changes made by this PR:
* Upgrades AWS-SDK to v1.38.61 and regenerates any mocks 
* Adds integration test for updating a managed nodegroup that has an `UpdateConfig`
* Validates that the nodegroup has an `UpdateConfig` before updating it
* Calls the AWS `UpdateNodegroupConfig` API to update the nodegroup's `UpdateConfig`

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

**Comment by Callisto13** [comment]
is it alright if i don't review this until the other one is merged? there is so much diff which i know is not part of this and i am not sure where the line is 😂

**Comment by nikimanoledaki** [comment]
> is it alright if i don't review this until the other one is merged? there is so much diff which i know is not part of this and i am not sure where the line is 😂

for sure, no worries! ☺️

File: integration/tests/managed/managed_nodegroup_test.go [hunk 1]
```diff
@@ -444,23 +444,18 @@ var _ = Describe("(Integration) Create Managed Nodegroups", func() {
...
+						UpdateConfig: &api.NodeGroupUpdateConfig{
+							MaxUnavailable: aws.Int(1),
+						},
 					},
 				}
-				cmd := params.EksctlUpdateCmd.
+				cmd = params.EksctlUpdateCmd.
 					WithArgs(
 						"nodegroup",
 						"--config-file", "-",
```
Comment (aclevername) [inline integration/tests/managed/managed_nodegroup_test.go:461]: do we want to assert it worked? e.g. fetching the new config from the EKS API to see its set to `1`? We do some direct EKS API calls in other integration tests (`unowned` for sure)

Comment (nikimanoledaki) [reply integration/tests/managed/managed_nodegroup_test.go:461]: ooo yeah good idea, let's do that! I'll add it!

File: pkg/actions/nodegroup/update.go [hunk 2]
```diff
@@ -3,26 +3,56 @@ package nodegroup
...
+		updateConfig.MaxUnavailablePercentage = aws.Int64(int64(*ng.UpdateConfig.MaxUnavailablePercentage))
+	}
+
+	_, err = m.ctl.Provider.EKS().UpdateNodegroupConfig(&eks.UpdateNodegroupConfigInput{
+		UpdateConfig:  updateConfig,
+		ClusterName:   &m.cfg.Metadata.Name,
+		NodegroupName: &ng.Name,
+	})
+	if err != nil {
+		return err
```
Comment (aclevername) [inline pkg/actions/nodegroup/update.go:53]: shall we wrap this error to provide more context?

Comment (nikimanoledaki) [reply pkg/actions/nodegroup/update.go]: done in https://github.com/weaveworks/eksctl/pull/3855/commits/2d5f9b706719cf0e24ed805b1d9e6148c7c41422 !

File: integration/tests/managed/managed_nodegroup_test.go (hunk 1 above)
Comment (nikimanoledaki) [reply integration/tests/managed/managed_nodegroup_test.go]: done in https://github.com/weaveworks/eksctl/pull/3855/commits/2d5f9b706719cf0e24ed805b1d9e6148c7c41422 !

File: pkg/actions/nodegroup/update.go (hunk 2 above)
Comment (nikimanoledaki) [reply pkg/actions/nodegroup/update.go]: yes good idea thanks, added this!

File: integration/tests/managed/managed_nodegroup_test.go [hunk 3]
```diff
@@ -444,23 +445,25 @@ var _ = Describe("(Integration) Create Managed Nodegroups", func() {
...
+				Expect(err).NotTo(HaveOccurred())
+				ctl := clusterProvider.Provider
+				out, err := ctl.EKS().DescribeNodegroup(&awseks.DescribeNodegroupInput{
+					ClusterName:   &params.ClusterName,
+					NodegroupName: aws.String("update-config-ng"),
+				})
+				Expect(err).NotTo(HaveOccurred())
+				Eventually(out.Nodegroup.UpdateConfig.MaxUnavailable).Should(Equal(aws.Int64(2)))
+
+				By("and upddating the nodegroup's UpdateConfig")
```
Comment (aclevername) [inline integration/tests/managed/managed_nodegroup_test.go:460]:
```suggestion
				By("and updating the nodegroup's UpdateConfig")
```

File: integration/tests/managed/managed_nodegroup_test.go [hunk 4]
```diff
@@ -472,6 +475,13 @@ var _ = Describe("(Integration) Create Managed Nodegroups", func() {
 				Expect(cmd).To(RunSuccessfullyWithOutputStringLines(
 					ContainElement(ContainSubstring("unchanged fields: the following fields remain unchanged; they are not supported by `eksctl update nodegroup`: Spot")),
 				))
+
+				out, err = ctl.EKS().DescribeNodegroup(&awseks.DescribeNodegroupInput{
+					ClusterName:   &params.ClusterName,
+					NodegroupName: aws.String("update-config-ng"),
+				})
+				Expect(err).NotTo(HaveOccurred())
+				Eventually(out.Nodegroup.UpdateConfig.MaxUnavailable).Should(Equal(aws.Int64(1)))
```
Comment (aclevername) [inline integration/tests/managed/managed_nodegroup_test.go:484]: nice!

File: pkg/actions/nodegroup/update.go [hunk 5]
```diff
@@ -3,26 +3,56 @@ package nodegroup
...
 	}
 
-	logger.Info("nodegroup successfully updated")
+	if ng.UpdateConfig == nil {
+		return fmt.Errorf("the submitted config didn't contain changes for nodegroup %s", ng.Name)
+	}
+
+	if describeNodegroupOutput.Nodegroup.UpdateConfig == nil {
+		return errors.New("cannot update updateConfig because the nodegroup is not configured to use one")
+	}
```
[the remark this answers is not in the document]
Comment (nikimanoledaki) [reply pkg/actions/nodegroup/update.go:34]: good point! the docs say that it defaults to `maxUnavailable=1` so I'll see if that's what's returned in the beta release on the `eksctl` account or if it returns `nil` ☺️

[the remark this answers is not in the document]
Comment (nikimanoledaki) [reply pkg/actions/nodegroup/update.go:34]: this took a bit longer than expected but you're right that it's never `nil`, even when `updateConfig` is not set during creation it always returns `maxUnavailable=1`. I checked this with a random old managed nodegroup that I found laying around and it also exposes `maxUnavailable=1`. will remove this, very good spot! 😄

File: pkg/actions/nodegroup/update.go [hunk 6]
```diff
@@ -3,26 +3,56 @@ package nodegroup
...
 )
 
 func (m *Manager) Update() error {
-	ngName := m.cfg.ManagedNodeGroups[0].Name
-	_, err := m.ctl.Provider.EKS().DescribeNodegroup(&eks.DescribeNodegroupInput{
+	ng := m.cfg.ManagedNodeGroups[0]
+	describeNodegroupOutput, err := m.ctl.Provider.EKS().DescribeNodegroup(&eks.DescribeNodegroupInput{
 		ClusterName:   &m.cfg.Metadata.Name,
-		NodegroupName: &ngName,
+		NodegroupName: &ng.Name,
```
[the remark this answers is not in the document]
Comment (nikimanoledaki) [reply pkg/actions/nodegroup/update.go:18]: good question - just tested it with a couple of nodegroups running k8s `v1.16`. the one with `maxUnavailable=2` returned that and the one without a set config returned the default `maxUnavailable=1` so backward compatibility seems good to me! so I don't think we need a version check :)

[the remark this answers is not in the document]
Comment (aclevername) [reply pkg/actions/nodegroup/update.go]: sounds good!

**Review by aclevername — APPROVED** [review]
LGTM!

**Pull request merged**